plugins/sharding/logical_count.rb

```diff
@@ -56,179 +41,134 @@ def cache_key(input)
-          @temporary_expressions = []
-          @referred_objects = []
+      class Counter < StreamExecutor
+        def initialize(context)
+          super(context, ShardCountExecutor)
         end
 
-        def count_pre(shard, shard_range)
-          cover_type = @target_range.cover_type(shard_range)
-          return if cover_type == :none
-
-          shard_key = shard.key
-          if shard_key.nil?
-            message = "[logical_count] shard_key doesn't exist: " +
-                      "<#{shard.key_name}>"
-            raise InvalidArgument, message
-          end
-
-          table_name = shard.table_name
-          range_index = nil
-          if @filter or @post_filter
-            log_use_range_index(false, table_name, "need filter",
-                                __LINE__, __method__)
-          elsif cover_type == :all
-            log_use_range_index(false, table_name, "covered",
-                                __LINE__, __method__)
-          else
-            index_info = shard_key.find_index(Operator::LESS)
-            range_index = index_info.index if index_info
-            if range_index
-              @referred_objects << range_index
-              log_use_range_index(true, table_name, "range index is available",
-                                  __LINE__, __method__)
-            else
-              log_use_range_index(false, table_name, "no range index",
-                                  __LINE__, __method__)
-            end
-          end
-          @contexts << ShardCountContext.new(shard, cover_type, range_index)
-        end
-
-        def count
-          prepare_contexts
+        def execute
+          have_shard = false
           total = 0
-          @contexts.each do |context|
-            total += count_shard(context)
+          each_shard_executor do |shard_executor|
+            have_shard = true
+            total += shard_executor.execute
+          end
+          unless have_shard
+            enumerator = @context.enumerator
+            message =
+              "[logical_count] no shard exists: " +
+              "logical_table: <#{enumerator.logical_table}>: " +
+              "shard_key: <#{enumerator.shard_key_name}>"
+            raise InvalidArgument, message
           end
           total
         end
+      end
+
+      class ShardCountExecutor < StreamShardExecutor
+        include Loggable
+        include QueryLoggable
 
-        def close
-          @temporary_expressions.each(&:close)
-          @temporary_tables.each(&:close)
-          @referred_objects.each(&:unref)
+        def initialize(context, shard, shard_range)
+          super("logical_count", context, shard, shard_range)
         end
 
-        private
-        def log_use_range_index(use, table_name, reason, line, method)
-          message = "[logical_count]"
-          if use
-            message << "[range-index]"
-          else
-            message << "[select]"
+        def execute
```

Comment:
It seems that most of this code similar to `LogicalRangeFilterCommand::ShardExecutor#execute`.
Can we unify it and this?

Reply:
Okay.
I have create a new PR #1346 to unity it and this.

Reply:
#1346 is merged